dwave/plugins/sklearn/utilities.py

```diff
@@ -224,3 +234,267 @@ def dot_2d(a: npt.ArrayLike, b: npt.ArrayLike, *,
+    ) -> npt.ArrayLike:
+    """
+    For the feature array `X` and the target array `y` computes
+    the matrix of (conditional) mutual information interactions. 
+    
+    cmi_{i, j} = I(x_i; y)
+    
+    If `conditional = True`, then the off-diagonal terms are computed:
+    
+    cmi_{i, j} = (I(x_i; y| x_j) + I(x_j; y| x_i)) / 2
+    
+    Otherwise
+    
+    cmi_{i, j} = I(x_i; x_j)
+    
+    Computation of I(x; y) uses the scikit-learn implementation, i.e.,
+    :func:`sklearn.feature_selection._mutual_info._estimate_mi`. The computation 
+    of I(x; y| z) is based on
+    https://github.com/jannisteunissen/mutual_information
+
+    Args:
+        X: See :func:`sklearn.feature_selection._mutual_info._estimate_mi`
+        
+        y: See :func:`sklearn.feature_selection._mutual_info._estimate_mi`
+        
+        conditional: bool, default=True
+            Whether to compute the off-diagonal terms using the conditional mutual
+            information or joint mutual information
+
+        discrete_features: See :func:`sklearn.feature_selection._mutual_info._estimate_mi`
+        
+        discrete_target: See :func:`sklearn.feature_selection._mutual_info._estimate_mi`
+        
+        n_neighbors: See :func:`sklearn.feature_selection._mutual_info._estimate_mi`
+        
+        copy: See :func:`sklearn.feature_selection._mutual_info._estimate_mi`
+        
+        random_state: See :func:`sklearn.feature_selection._mutual_info._estimate_mi`
+
+        n_workers: int, default=1
+            Number of workers for parallel computation on the cpu       
+        
+    Returns:
+        mi_matrix : ndarray, shape (n_features, n_features)
+        Interaction matrix between the features using (conditional) mutual information.
+        A negative value will be replaced by 0.
+
+    References:    
+    .. [1] A. Kraskov, H. Stogbauer and P. Grassberger, "Estimating mutual
+           information". Phys. Rev. E 69, 2004.
+    .. [2] B. C. Ross "Mutual Information between Discrete and Continuous
+           Data Sets". PLoS ONE 9(2), 2014.
+    .. [3] Mesner, Octavio César, and Cosma Rohilla Shalizi. "Conditional
+           mutual information estimation for mixed, discrete and continuous
+           data." IEEE Transactions on Information Theory 67.1 (2020): 464-484.
+    """
+    
+    X, y = check_X_y(X, y, accept_sparse="csc", y_numeric=not discrete_target)
+    n_samples, n_features = X.shape
+    
+    if isinstance(discrete_features, (str, bool)):
+        if isinstance(discrete_features, str):
+            if discrete_features == "auto":
+                discrete_features = issparse(X)
+            else:
+                raise ValueError("Invalid string value for discrete_features.")
+        discrete_mask = np.empty(n_features, dtype=bool)
+        discrete_mask.fill(discrete_features)
+    else:
+        discrete_features = check_array(discrete_features, ensure_2d=False)
+        if discrete_features.dtype != "bool":
```

Comment:
Much better to use something like https://numpy.org/devdocs/reference/generated/numpy.issubdtype.html